File: packages/kindling-core/src/retrieval/orchestrator.ts
/**
 * Retrieval orchestrator
 *
 * Combines pins, summaries, and provider candidates into a unified retrieval response.
 */

import type {
  RetrieveOptions,
  RetrieveResult,
  PinResult,
  CandidateResult,
  RetrievalProvider,
} from '../types/retrieval.js';
import type { Pin, Observation, Summary, ID } from '../types/index.js';

/**
 * Store interface for retrieval operations
 */
export interface RetrievalStore {
  /**
   * List active pins for a scope
   */
  listActivePins(scopeIds?: Partial<Record<string, string>>, now?: number): Pin[];

  /**
   * Get observation by ID
   */
  getObservationById(observationId: ID): Observation | undefined;

  /**
   * Get summary by ID
   */
  getSummaryById(summaryId: ID): Summary | undefined;

  /**
   * Get open capsule for session
   */
  getOpenCapsuleForSession(sessionId: string): { id: ID; summaryId?: ID } | undefined;

  /**
   * Get latest summary for capsule
   */
  getLatestSummaryForCapsule(capsuleId: ID): Summary | undefined;
}

/**
 * Retrieve relevant context for a query
 *
 * Orchestrates retrieval by combining:
 * 1. Active pins (non-evictable, always included)
 * 2. Current session summary (non-evictable if exists)
 * 3. Provider candidates (ranked, budget-limited)
 *
 * @param store - Retrieval store
 * @param provider - Retrieval provider
 * @param options - Retrieval options
 * @returns Retrieval result with pins, summary, and candidates
 */
export async function retrieve(
  store: RetrievalStore,
  provider: RetrievalProvider,
  options: RetrieveOptions,
): Promise<RetrieveResult> {
  const { query, scopeIds, maxCandidates = 10, includeRedacted = false } = options;

  const now = Date.now();

  // Step 1: Fetch active pins for scope
  const pins = store.listActivePins(scopeIds as Partial<Record<string, string>>, now);

  // Step 2: Resolve pins to their targets
  const pinResults: PinResult[] = [];
  const pinnedIds = new Set<ID>();

  for (const pin of pins) {
    let target: Observation | Summary | undefined;

    if (pin.targetType === 'observation') {
      target = store.getObservationById(pin.targetId);
    } else if (pin.targetType === 'summary') {
      target = store.getSummaryById(pin.targetId);
    }

    if (target) {
      // Skip redacted unless explicitly requested
      if ('redacted' in target && target.redacted && !includeRedacted) {
        continue;
      }

      pinResults.push({ pin, target });
      pinnedIds.add(target.id);
    }
  }

  // Step 3: Get current session summary (non-evictable)
  let currentSummary: Summary | undefined;

  if (scopeIds.sessionId) {
    const capsule = store.getOpenCapsuleForSession(scopeIds.sessionId);
    if (capsule) {
      const summary = store.getLatestSummaryForCapsule(capsule.id);
      if (summary) {
        currentSummary = summary;
        pinnedIds.add(summary.id);
      }
    }
  }

  // Step 4: Get provider candidates (exclude pinned and current summary)
  const providerResults = await provider.search({
    query,
    scopeIds,
    maxResults: maxCandidates,
    excludeIds: Array.from(pinnedIds),
    includeRedacted,
  });

  // Step 5: Convert provider results to candidates
  const candidates: CandidateResult[] = providerResults.map((result) => ({
    entity: result.entity,
    score: result.score,
    matchContext: result.matchContext,
  }));

  // Step 6: Build provenance
  const provenance = {
    query,
    scopeIds,
    totalCandidates: providerResults.length,
    returnedCandidates: candidates.length,
    truncatedDueToTokenBudget: false, // Token budgeting not implemented in v0.1
    providerUsed: provider.name,
  };

  return {
    pins: pinResults,
    currentSummary,
    candidates,
    provenance,
  };
}

File: crates/kindling-provider/src/orchestrator.rs
//! Retrieval orchestrator.
//!
//! Combines pins, the current session summary, and provider candidates into a
//! unified [`RetrieveResult`]. Mirrors `retrieve` in
//! `packages/kindling-core/src/retrieval/orchestrator.ts`:
//!
//! 1. Active pins (non-evictable, always included)
//! 2. Current session summary (non-evictable if it exists)
//! 3. Provider candidates (ranked, capped at `max_candidates`)
//!
//! Token-budget assembly is deliberately absent: `token_budget` is deprecated
//! and budgeting is a downstream-system responsibility.

use std::time::{SystemTime, UNIX_EPOCH};

use kindling_store::SqliteKindlingStore;
use kindling_types::{
    CandidateResult, Id, PinResult, PinTargetType, RetrieveOptions, RetrieveProvenance,
    RetrieveResult, RetrievedEntity, Timestamp,
};

use crate::error::ProviderResult;
use crate::provider::RetrievalProvider;

/// Default candidate cap when `max_candidates` is not given.
const DEFAULT_MAX_CANDIDATES: u32 = 10;

/// Retrieve relevant context for a query, scored as of the current time.
pub fn retrieve(
    store: &SqliteKindlingStore,
    provider: &impl RetrievalProvider,
    options: &RetrieveOptions,
) -> ProviderResult<RetrieveResult> {
    retrieve_at(store, provider, options, now_ms())
}

/// Retrieve relevant context for a query, scored as of `now` (epoch ms).
///
/// The explicit clock is the only deviation from the TS orchestrator (which
/// reads `Date.now()`); it keeps pin expiry and recency scoring deterministic.
pub fn retrieve_at(
    store: &SqliteKindlingStore,
    provider: &impl RetrievalProvider,
    options: &RetrieveOptions,
    now: Timestamp,
) -> ProviderResult<RetrieveResult> {
    let max_candidates = options.max_candidates.unwrap_or(DEFAULT_MAX_CANDIDATES);
    let include_redacted = options.include_redacted.unwrap_or(false);

    // Step 1: active pins for the scope.
    let pins = store.list_active_pins(Some(&options.scope_ids), Some(now))?;

    // Step 2: resolve pins to their targets.
    let mut pin_results: Vec<PinResult> = Vec::new();
    let mut pinned_ids: Vec<Id> = Vec::new();

    for pin in pins {
        let target = match pin.target_type {
            PinTargetType::Observation => store
                .get_observation_by_id(&pin.target_id)?
                .map(RetrievedEntity::Observation),
            PinTargetType::Summary => store
                .get_summary_by_id(&pin.target_id)?
                .map(RetrievedEntity::Summary),
        };

        if let Some(target) = target {
            // Skip redacted observations unless explicitly requested.
            if let RetrievedEntity::Observation(obs) = &target {
                if obs.redacted && !include_redacted {
                    continue;
                }
            }
            let target_id = entity_id(&target).to_string();
            if !pinned_ids.contains(&target_id) {
                pinned_ids.push(target_id);
            }
            pin_results.push(PinResult { pin, target });
        }
    }

    // Step 3: current session summary (non-evictable).
    // The TS orchestrator gates on JS truthiness, so an empty-string session
    // ID is treated as absent here too.
    let mut current_summary = None;
    if let Some(session_id) = options
        .scope_ids
        .session_id
        .as_deref()
        .filter(|s| !s.is_empty())
    {
        if let Some(capsule) = store.get_open_capsule_for_session(session_id)? {
            if let Some(summary) = store.get_latest_summary_for_capsule(&capsule.id)? {
                if !pinned_ids.contains(&summary.id) {
                    pinned_ids.push(summary.id.clone());
                }
                current_summary = Some(summary);
            }
        }
    }

    // Step 4: provider candidates, excluding pinned IDs and the current summary.
    let provider_results = provider.search(
        &kindling_types::ProviderSearchOptions {
            query: options.query.clone(),
            scope_ids: options.scope_ids.clone(),
            max_results: Some(max_candidates),
            exclude_ids: Some(pinned_ids),
            include_redacted: Some(include_redacted),
        },
        now,
    )?;

    // Step 5: convert provider results to candidates.
    let total_candidates = provider_results.len() as u32;
    let candidates: Vec<CandidateResult> = provider_results
        .into_iter()
        .map(|result| CandidateResult {
            entity: result.entity,
            score: result.score,
            match_context: result.match_context,
        })
        .collect();

    // Step 6: provenance.
    let provenance = RetrieveProvenance {
        query: options.query.clone(),
        scope_ids: options.scope_ids.clone(),
        total_candidates,
        returned_candidates: candidates.len() as u32,
        truncated_due_to_token_budget: false,
        provider_used: provider.name().to_string(),
    };

    Ok(RetrieveResult {
        pins: pin_results,
        current_summary,
        candidates,
        provenance,
    })
}

fn entity_id(entity: &RetrievedEntity) -> &str {
    match entity {
        RetrievedEntity::Observation(obs) => &obs.id,
        RetrievedEntity::Summary(sum) => &sum.id,
    }
}

fn now_ms() -> Timestamp {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system clock before Unix epoch")
        .as_millis() as Timestamp
}
